In [1]:
import sys
from imp import reload
sys.path.insert(0, "/media/diskD/EPFL/Fall 2016/ADA/Homework/Gokcen_Meryem_Riyadh_ADA/03 - Interactive Viz")
reload(sys)

from create_map import *
from university_canton_mappings import *
import pandas as pd
import html5lib
import folium
import numpy as np
import googlemaps
from datetime import datetime
import json

# Exercise 1

The sole purpose of this exercise is not only to calculate the total amount of grants alotted to different Swiss cantons but it is about creating a visualization that shows those differences in an interactive way through the use of choropleth maps.

### Step 1: Extraction of Data from Grants csv file

We work with P3 database of the SNSF (Swiss National Science Foundation) to extract grant information. The csv file that conveys in particular our purpose is P3_GrantExport.csv which we downloaded and put under Data directory.  

Our strategy in this part consists of the following:
1. Data Extraction: after familiarization with schema, we keep only useful columns
2. Data Cleaning: removal of rows for which the university/ totalAmount are not provided  
3. Data Pre-Processing: convertion of totalAmounts to float type so that the aggregation by the sum of total amounts per each university works

#### Step 1.1. Data Extraction:

We start by loading the data and looking at the first 5 rows.  

In [3]:
original_grants = pd.read_csv('Data/P3_GrantExport.csv',sep=";")
original_grants.head()

"Project Number"                                      Project Title  \
0                  1  Schlussband (Bd. VI) der Jacob Burckhardt-Biog...   
1                  4  Batterie de tests à l'usage des enseignants po...   
2                  5  Kritische Erstausgabe der "Evidentiae contra D...   
3                  6  Katalog der datierten Handschriften in der Sch...   
4                  7  Wissenschaftliche Mitarbeit am Thesaurus Lingu...   

  Project Title English                              Responsible Applicant  \
0                   NaN                                       Kaegi Werner   
1                   NaN                                 Massarenti Léonard   
2                   NaN  Kommission für das Corpus philosophorum medii ...   
3                   NaN                                     Burckhardt Max   
4                   NaN                       Schweiz. Thesauruskommission   

             Funding Instrument Funding Instrument Hierarchy  \
0  Project funding (Div. I-III)              Project funding   
1  Project funding (Div. I-III)              Project funding   
2  Project funding (Div. I-III)              Project funding   
3  Project funding (Div. I-III)              Project funding   
4  Project funding (Div. I-III)              Project funding   

                                         Institution  \
0                                                NaN   
1  Faculté de Psychologie et des Sciences de l'Ed...   
2  Kommission für das Corpus philosophorum medii ...   
3  Abt. Handschriften und Alte Drucke Bibliothek ...   
4                       Schweiz. Thesauruskommission   

                                University  Discipline Number  \
0                     Nicht zuteilbar - NA              10302   
1                Université de Genève - GE              10104   
2  NPO (Biblioth., Museen, Verwalt.) - NPO              10101   
3                   Universität Basel - BS              10302   
4  NPO (Biblioth., Museen, Verwalt.) - NPO              10303   

                         Discipline Name  \
0                          Swiss history   
1       Educational science and Pedagogy   
2                             Philosophy   
3                          Swiss history   
4  Ancient history and Classical studies   

                           Discipline Name Hierarchy  Start Date    End Date  \
0  Human and Social Sciences;Theology & religious...  01.10.1975  30.09.1976   
1  Human and Social Sciences;Psychology, educatio...  01.10.1975  30.09.1976   
2  Human and Social Sciences;Linguistics and lite...  01.03.1976  28.02.1985   
3  Human and Social Sciences;Theology & religious...  01.10.1975  30.09.1976   
4  Human and Social Sciences;Theology & religious...  01.01.1976  30.04.1978   

  Approved Amount Keywords  
0        11619.00      NaN  
1        41022.00      NaN  
2        79732.00      NaN  
3        52627.00      NaN  
4       120042.00      NaN

We keep only two columns: University and Approved Amount

In [4]:
amount_university_grants = original_grants[["University","Approved Amount"]]

#### Step 1.2. Data Cleaning:

In [5]:
amount_university_grants["Approved Amount"].describe()

count                       63969
unique                      36455
top       data not included in P3
freq                        10910
Name: Approved Amount, dtype: object

In [6]:
len(amount_university_grants[amount_university_grants["Approved Amount"]=='data not included in P3'])

10910

For 10910 rows, the approved amount is not included and have a value of 'data not included in P3'. We decided to remove those. 

In [7]:
grants_without_nan = amount_university_grants[amount_university_grants["Approved Amount"]!='data not included in P3']

In [8]:
grants_without_nan.describe()

University Approved Amount
count                     50878           53059
unique                       77           36454
top     Universität Zürich - ZH        10000.00
freq                       6754             611

#### 3. Data Pre-Processing:

In [9]:
grants_without_nan["Approved Amount"].dtype

dtype('O')

Since the type of this column is object, we convert it into numeric values.

In [10]:
grants_without_nan["Approved Amount"] = pd.to_numeric(grants_without_nan["Approved Amount"],errors='coerce')
grants_without_nan

/home/meryem/miniconda2/envs/py3k/lib/python3.5/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


University  Approved Amount
0                                   Nicht zuteilbar - NA          11619.0
1                              Université de Genève - GE          41022.0
2                NPO (Biblioth., Museen, Verwalt.) - NPO          79732.0
3                                 Universität Basel - BS          52627.0
4                NPO (Biblioth., Museen, Verwalt.) - NPO         120042.0
5                            Université de Fribourg - FR          53009.0
6                            Université de Fribourg - FR          25403.0
7                                Universität Zürich - ZH          47100.0
8                            Université de Lausanne - LA          25814.0
9                              Université de Genève - GE         360000.0
10                           Université de Fribourg - FR         153886.0
11               NPO (Biblioth., Museen, Verwalt.) - NPO         862200.0
12                             Université de Genève - GE         116991.0
13                                Universität Basel - BS         112664.0
14                               Universität Zürich - ZH           5000.0
15                           Université de Fribourg - FR         204018.0
16                             Université de Genève - GE         149485.0
17                               Universität Zürich - ZH          83983.0
18                                  Nicht zuteilbar - NA          38152.0
19                           Université de Lausanne - LA          14138.0
20                             Université de Genève - GE         164602.0
21                                 Universität Bern - BE         147795.0
22                               Universität Zürich - ZH          24552.0
23                             Université de Genève - GE          44802.0
24                                  Nicht zuteilbar - NA          56000.0
25                                  Nicht zuteilbar - NA         152535.0
26               NPO (Biblioth., Museen, Verwalt.) - NPO         225000.0
27               NPO (Biblioth., Museen, Verwalt.) - NPO         179124.0
28                               Universität Zürich - ZH          20000.0
29     Eidg. Forschungsanstalt für Wald,Schnee,Land -...         445198.0
...                                                  ...              ...
63939                           Hochschule Luzern - HSLU           1898.0
63940                                                NaN           9300.0
63941                                                NaN           9500.0
63942                                                NaN           4000.0
63943                                                NaN          11000.0
63944                                                NaN           7500.0
63945                                                NaN           8000.0
63946                                                NaN           9300.0
63947                                                NaN           7000.0
63948                                                NaN           6000.0
63949                                                NaN           4500.0
63950            NPO (Biblioth., Museen, Verwalt.) - NPO           2800.0
63951                                  ETH Zürich - ETHZ          99952.0
63952                Forschungsanstalten Agroscope - AGS          98040.0
63953                       Firmen/Privatwirtschaft - FP         100000.0
63954                                                NaN          10500.0
63955                             Universität Basel - BS          10030.0
63956                                                NaN           7870.0
63957                                                NaN           3150.0
63958                            Universität Zürich - ZH          10600.0
63959                        Université de Lausanne - LA           2515.0
63960                                                NaN           3500.0
63961                                             

In [11]:
grants_without_nan.dtypes

University          object
Approved Amount    float64
dtype: object

We aggregate the dataframe by calculating the sum of approved amount per each university.

In [12]:
df = grants_without_nan.groupby(["University"])["Approved Amount"].sum().reset_index()

In [13]:
grants_without_nan.dtypes

University          object
Approved Amount    float64
dtype: object

We convert it into a dataframe

In [14]:
university_amount_df = pd.DataFrame(df)

University  Approved Amount
0                       AO Research Institute - AORI        3435621.0
1               Allergie- und Asthmaforschung - SIAF       19169965.0
2                        Berner Fachhochschule - BFH       31028695.0
3             Biotechnologie Institut Thurgau - BITG        2492535.0
4  Centre de rech. sur l'environnement alpin - CR...        1567678.0

#### Data Description Summary:

In [ ]:
print('Number of rows after filtering data: ',len(grants_without_nan),
      'Ratio of kept rows after filtering the data: ',(len(grants_without_nan)/len(original_grants))*100
      'Number of universities:',len(university_amount_df))

### Step 2: Mapping University names to cantons

Now that we have compiled the list of universities, we move to the next step which is mapping universities to cantons which is accomplished in two steps:
1. Automatic Mapping: we have tried different APIs including geomaps, google maps. But at the end, we opted for google maps api thanks to its efficiency and its exhaustivity of the returned results. We also played a little bit with the universities names to maximize the coverage: we compare the performance using first part of the university name versus its abbreviation. We got better results by splitting the university name and taking only the first part without the abbreviation.  We have also tried goslate and other translation apis to translate to english but it didn't increase the performance of the coverage so much. At the end, the best coverage we got is 35 out of 77 universities which is roughly half coverage. 
2. Manual Mapping: for those universities for which we couldn't get a mapping automatically, we looked them up manually and we compiled this list in a separate file called manual_mappings under Data directory. 

#### Step 2.1: Mapping using GoogleMaps API

We add the component 'components=country:ch' to force the api to look only in Switzerland

#### Remark:

For safety reasons, we don't include the key we used. If you want to try it, please use your own key by assigning a value to key.

In [16]:
key = # Add you key here
google_url = 'https://maps.googleapis.com/maps/api/geocode/json?components=country:ch&key='+key
canton_mappings=[]
for index, row in university_amount_df.iterrows():
    university = row['University'].split('-')[0]
    canton=get_canton_id(google_url,university)
    canton_mappings.append(canton)

ConnectionError: HTTPSConnectionPool(host='maps.googleapis.com', port=443): Max retries exceeded with url: /maps/api/geocode/json?components=country:ch&key=AIzaSyDQMfP1in42ay64Ont3UduALKwYhmWeh90&address=AO+Research+Institute+ (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7fcded5ad2e8>: Failed to establish a new connection: [Errno -3] Temporary failure in name resolution',))

In [ ]:
university_amount_df['Canton'] = canton_mappings

#### Summary of automatic mapping

In [ ]:
canton_mappings

In [ ]:
university_mapped = university_amount_df[university_amount_df['Canton']!='']
university_total = university_amount_df
print('Total Number of universities:'+len(canton_mappings),
      'Total Number of universities for which canton was found:'+len(university_mapped)/len(university_total))

### Step 3: Manual Mappings

In [ ]:
# Reading manual mapping data from file and storing them in a dataframe
manual_mapping_df = pd.read_csv('Data/manualMapping.txt',header=None)
manual_mapping_df.columns = ['University','Canton']
manual_mapping_df
for i in range(0,len(manual_mapping_df)):
    university_amount_df['Canton'][university_amount_df['University']==manual_mapping_df['University'][i]]=manual_mapping_df['Canton'][i]

In [ ]:
# Displaying the whole entries of universities
university_amount_df

### Step 4:  Final Preparation of Mappings

#### Step 4.1: Filtering Out and Testing for University/Amount coverage

In [ ]:
# Display the entries of universities for which we get empty results of cantons
university_amount_df_nan = university_amount_df[university_amount_df['Canton']=='']
university_amount_df_nan

In [ ]:
university_amount_df_nan.describe()

There are three categories of universities for which we didn't succeed to find a mapping:
1. Universities provided under the name of "Other Universities" as translated from german:
   	'Weitere Institute - FINST' and 'Weitere Spitäler - ASPIT', '	NPO (Biblioth., Museen, Verwalt.) - NPO'
2. No assignable values: 'Nicht zuteilbar - NA'
3. Universities not in Switzerland: '7 Istituto Svizzero di Roma - ISR' which is in Roma,    Italy
4. Universities we found in multiple cantons: 'Eidg. Forschungsanstalt für Wald,Schnee,Land - WSL' and 'Eidg. Material und Prüfungsanstalt - EMPA'
5. Private Companies: 'Firmen/Privatwirtschaft - FP'

In [ ]:
# Dropping Italian University
swiss_university_amount_df = university_amount_df[university_amount_df['University']!="Istituto Svizzero di Roma - ISR"]
# Dropping Private Companies
swiss_university_amount_df = swiss_university_amount_df[swiss_university_amount_df['University']!="Firmen/Privatwirtschaft - FP"]
swiss_university_amount_df.reset_index(drop=True)

In [ ]:
sum_all_df = swiss_university_amount_df['Approved Amount'].sum()
sum_nan_df = swiss_university_amount_df[swiss_university_amount_df["Canton"]=='']['Approved Amount'].sum()
percent = sum_nan_df/sum_all_df 
percent

The percentage of amounts not covered is 4.69%, so it is safe to say that we have covered top 95% Swiss universities.

In [ ]:
university_amount_df = university_amount_df[university_amount_df["Canton"]!='']
university_amount_df

In [ ]:
university_amount_df

#### Step 4.2: Total Amount per canton Mappings Final Stage

Now, we aggregate using groupby to get the sum of the amounts for each canton.

In [ ]:
df = university_amount_df.groupby(["Canton"])["Approved Amount"].sum().reset_index()

In [ ]:
canton_amount_df = pd.DataFrame(df)
canton_amount_df

#### Step 4.3: Getting the cantons from json file and assigning 0 to cantons to which no university in our database maps to

In [ ]:
# Do not forget to explain the format of the json
cantons_mappings_dict = get_cantons_json('json/ch-cantons.topojson.json')
cantons_mappings_dict.values()

In [ ]:
existing_cantons = canton_amount_df['Canton'].values
for canton_id in cantons_mappings_dict.values():
    if (canton_id not in existing_cantons):
        canton_amount_df.loc[len(canton_amount_df)]=[canton_id,0]
        
canton_amount_df

### Step 5: Binding of data between Pandas DataFrame and Geo/TopoJSON geometries and Creation of Choropleth Map

To explain the coloring threshold we applied to the map ....

In [ ]:
amounts_df =canton_amount_df.sort('Approved Amount')
amounts_df.reset_index()
amounts_df['Delta']=(amounts_df['Approved Amount'] - amounts_df['Approved Amount'].shift(1))
amounts_df['Delta_x']= amounts_df['Delta']/1000000
amounts_df['cum_sum'] = amounts_df['Approved Amount'].cumsum()
amounts_df['cum_sum_x']= amounts_df['cum_sum']/1000000
amounts_df

Finally, we create our choropleth map using the function create_choropleth_map. Please refer to create_map.py python file for details of implementation of this function and further documentation.

In [1]:
swiss_map = create_choropleth_map(canton_amount_df)
swiss_map.save('html/swiss_map.html')
swiss_map

NameError: name 'create_choropleth_map' is not defined

# Bonus Part

We got Canton-Language table directly from wikipedia and use it.

In [7]:
table = pd.read_html('html/canton_language.html', header=0)[1][['Code','Official languages']]
table = table[:-1].sort_values('Code').reset_index(drop=True)
table

Code        Official languages
0    AG                    German
1    AI                    German
2    AR                    German
3    BE            German, French
4    BL                    German
5    BS                    German
6    FR            French, German
7    GE                    French
8    GL                    German
9    GR  German, Romansh, Italian
10   JU                    French
11   LU                    German
12   NE                    French
13   NW                    German
14   OW                    German
15   SG                    German
16   SH                    German
17   SO                    German
18   SZ                    German
19   TG                    German
20   TI                   Italian
21   UR                    German
22   VD                    French
23   VS            French, German
24   ZG                    German
25   ZH                    German

After doing manual inspection on problematic cantons [Bern, GR, Valais, Fribourg], we have concluded that:
Fribourg's amount can be divived in to two with %50 %50 ratios. It has two universities and they both teach both of the languages (French and German). Also, geographically, they are on the border of language seperation.
Universities at Bern are mostly located in Bern's center which is inside German part. So we assign 'German' to Bern.
Similiarly, in Graübunden, almost all universities there belong to German side
Finally, in Valais, schools that share the huge majority of the resources belong to French side, therefore we assign 'French' label.

To handle Fribourg's extreme case, we first assign it to French.

In [ ]:
table.loc[3]['Official languages'] = 'German' #Bern
table.loc[9]['Official languages'] = 'German' #GR
table.loc[6]['Official languages'] = 'French' #Fribourg
table.loc[23]['Official languages'] = 'French' #Valais

fribourg_amount = int(canton_amount_df.loc[6]['Approved Amount'])

merged = pd.merge(canton_amount_df, table, right_on='Code', left_on='Canton1')[['Approved Amount', 'Official languages']]

cantons_amount = merged.groupby(['Official languages']).sum().reset_index()


cantons_amount.set_value(0, 'Approved Amount', cantons_amount.loc[0]['Approved Amount'] - (fribourg_amount*0.5))
cantons_amount.set_value(1, 'Approved Amount', cantons_amount.loc[1]['Approved Amount'] + (fribourg_amount*0.5))
cantons_amount = cantons_amount.set_value(4, 'Official languages', 'Rumantsch')
cantons_amount = cantons_amount.set_value(4, 'Approved Amount', 0)

%matplotlib inline

plot = cantons_amount.set_index('Official languages').plot.pie(y='Approved Amount', x='Official languages')
plot.legend(loc='lower right')
plot.axis('equal')
plot